syntax = "proto3";

package books_database;

service BooksDatabaseService {
    rpc Prepare(GenericBookRequest) returns (PrepareResponse);
    rpc Commit(GenericBookRequest) returns (CommitResponse);
    rpc Abort(GenericBookRequest) returns (AbortResponse);

    rpc AddOrderLineToPrepared(GenericBookRequest) returns (WriteResponse);
    rpc RemoveOrderLineFromPrepared(GenericBookRequest) returns (WriteResponse);

    rpc Read(ReadRequest) returns (ReadResponse);
    rpc Write(WriteRequest) returns (WriteResponse);
}

message GenericBookRequest {
    string order_id = 1;
    string title = 2;
    int32 amount = 3;
}

message PrepareResponse {
    bool ready = 1;
    string message = 2;
}

message CommitResponse {
    bool success = 1;
    string message = 2;
}

message AbortResponse {
    bool aborted = 1;
    string message = 2;
}

message ReadRequest {
    string title = 1;
}

message ReadResponse {
    int32 stock = 1;
}

message WriteRequest {
    string title = 1;
    int32 new_stock = 2;
}

message WriteResponse {
    bool is_success = 1;
}
